File: backend/app/schemas/digest.py
# Copyright (C) 2026 Lorenzo Benfeati
# SPDX-License-Identifier: AGPL-3.0-or-later
import uuid
from datetime import datetime, timedelta

from pydantic import BaseModel, ConfigDict, Field, model_validator


class DigestResponse(BaseModel):
    id: uuid.UUID
    title: str | None
    period_start: datetime
    period_end: datetime
    content_html: str | None
    content_text: str | None
    virtual_feed_id: uuid.UUID | None
    llm_provider: str | None
    llm_model: str | None
    article_count: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class DigestGenerateOptions(BaseModel):
    period_start: datetime | None = None
    period_end: datetime | None = None
    virtual_feed_id: uuid.UUID | None = None
    category_ids: list[uuid.UUID] = Field(default_factory=list)
    max_articles: int = Field(default=30, ge=5, le=100)
    force_fulltext: bool = True
    llm_config_id: uuid.UUID | None = None

    @model_validator(mode="after")
    def set_defaults(self) -> "DigestGenerateOptions":
        now = datetime.utcnow()
        if self.period_end is None:
            self.period_end = now
        if self.period_start is None:
            self.period_start = now - timedelta(hours=24)
        if self.period_start >= self.period_end:
            raise ValueError("period_start must be before period_end")
        return self


class DigestListResponse(BaseModel):
    items: list[DigestResponse]
    total: int
    page: int
    pages: int

File: frontend/src/api/digest.ts
// Copyright (C) 2026 Lorenzo Benfeati
// SPDX-License-Identifier: AGPL-3.0-or-later
import client from './client'

export interface Digest {
  id: string
  title: string | null
  period_start: string
  period_end: string
  content_html: string | null
  content_text: string | null
  virtual_feed_id: string | null
  llm_provider: string | null
  llm_model: string | null
  article_count: number
  status: string
  generation_error: string | null
  created_at: string
}

export interface DigestGenerateOptions {
  period_start?: string
  period_end?: string
  virtual_feed_id?: string
  category_ids?: string[]
  max_articles?: number
  force_fulltext?: boolean
}

export const digestApi = {
  list: (params?: { page?: number; size?: number; virtual_feed_id?: string }) =>
    client.get<{ items: Digest[]; total: number; page: number; pages: number }>('/digests', { params }),

  get: (id: string) =>
    client.get<Digest>(`/digests/${id}`),

  latest: (virtualFeedId?: string) =>
    client.get<Digest>('/digests/latest', {
      params: virtualFeedId ? { virtual_feed_id: virtualFeedId } : {},
    }),

  generate: (options: DigestGenerateOptions = {}) =>
    client.post<Digest>('/digests/generate', options),

  delete: (id: string) =>
    client.delete(`/digests/${id}`),
}
